### DISCUSSION TITLE
How to configure a custom span processor on top of python auto-instrumentation

### DISCUSSION THREAD

**Discussion by bjrara** [open-telemetry/opentelemetry-python-contrib#1705] [General]
My request is very similar to the one asked in Java: [Best way to add a custom span processor on top of auto-configuration](https://github.com/open-telemetry/opentelemetry-java/discussions/4595), but this is in Python's version.

I have a custom span processor that adds additional attributes into spans based on the parent context. Is there any way to add it without user invoking `add_span_processor` in their codes?

Thanks!

**Comment by srikanthccv** [comment]
> Is there any way to add it without user invoking add_span_processor in their codes?


No, this cannot be done at the moment, but it's also not that difficult to add this capability. Add an [entry point](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/pyproject.toml) for the span processor with the default being the "[batch](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L132)" and let the user configure the span processor of their choice to be used in auto instrumentation like custom exporter [here](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_configuration/__init__.py#LL208C6-L208C6)

**Comment by bjrara** [comment]
Thanks @srikanthccv for the confirmation, and the code sharing.

If users don't use auto-instrumentation SDK, we can provide customized tracer provider and ask users to configure it using `OTEL_PYTHON_TRACER_PROVIDER` env var. However, according to the [link](https://github.com/open-telemetry/opentelemetry-python/blob/f05fa772e65b72a1fdc0a40e142e9342dee013f3/opentelemetry-sdk/src/opentelemetry/sdk/_configuration/__init__.py#L192) you sent to me, when using auto-instrumentation, this env var is not respected, and the tracer provider defaults to `opentelemetry.sdk.trace.TracerProvider`.

I'm wondering if this is discussed before whether to allow customers to wire a different tracer provider into auto-instrumentation. In Java auto-instrumentation, this feature is supported.

**Reply by kokikathir** [reply]
@bjrara Have you found any solution to add a custom processor with auto instrumentation?

**Comment by beaugunderson** [comment]
I've just spent about 12 hours of my life on this... is this possible yet? it does not seem so... I've tried to pre-create a `TracerProvider` and add the processor to that, so `initialize()` will use it instead of creating its own, that doesn't work... I've tried to get the `TracerProvider` created via `initialize()` and add it to that, that doesn't work... I've tried to make my own configurator and pass that in `OTEL_PYTHON_CONFIGURATOR`, that doesn't work (its `_configure` method is never called and I get a log message about how it won't be run because it doesn't mach the entry point name)

**Comment by yrosemacbs** [comment]
Also got here after trying in vain. For reference here's my use case:
We schedule jobs using [rq](https://github.com/rq/rq/) workers. The workers fork the proces so we can't use the batch span processor for tracing the jobs. As far as i can tell rq has no "post fork" method I can add a tracer to so I'd like to add the simple span processor to the existing tracer provider. I'd also like to rely on auto-instrumentation for the rest of the configuration so i can configure the exporter config using the kubernetes instrumentation resource. The problem is that I cannot get a hold of the configured exporter to add to my span processor. So i have to make assumptions and lock my code in with my infrastructure config.

The best solution I found is to implement a conditional import and exporter assignment, but that feels clunky..
